### PR TITLE
Split plugin running

### PR DESCRIPTION
**Pull request by mfleader**
## Changes introduced with this PR

- [x] Split the plugin provider's Running Stage into a Starting Stage and a Running Stage.
- [x] Add unit tests

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dustinblack** [comment]
It looks like your 'remove dead code' commit may have left some artifacts that need cleaning up for the CI (or perhaps that code wasn't quite dead yet?).

File: internal/step/plugin/provider.go [hunk 1]
```diff
@@ -85,6 +93,8 @@ const (
 	StageIDOutput StageID = "outputs"
 	// StageIDCrashed is a stage that indicates that the plugin has quit unexpectedly.
 	StageIDCrashed StageID = "crashed"
+	// StageIDStarting is a stage that indicates that the plugin execution has begun.
+	StageIDStarting StageID = "starting"
 )
 
 var deployingLifecycleStage = step.LifecycleStage{
```
Comment (Harshith-umesh) [inline internal/step/plugin/provider.go:100]: shouldn't the NextStages for the StageIDDeploy be StageIDStarting?

Comment (jaredoconnell) [reply internal/step/plugin/provider.go:100]: I think so, good catch!

**Review by jaredoconnell — CHANGES_REQUESTED** [review]
Looks good overall. I agree with a lot of your design choices.

My test workflows worked properly with this branch. Some of my other changes tainted the environment making it so I can't run some test cases. 

The only things that need fixing are the NextStages issue, and comments on basically all of the test cases that explain the high level goal of the test case, and the sections in the larger test cases. These comments will make it easier to review.

File: internal/step/plugin/provider.go (hunk 1 above)
Comment (dustinblack) [reply internal/step/plugin/provider.go:100]: It looks like maybe this was resolved when main was merged in?

File: internal/step/plugin/provider_test.go [hunk 2]
```diff
@@ -1 +1,383 @@
...
+	}
+	if previousStageOutput == nil {
+		panic(fmt.Errorf("no previous stage output ID"))
+	}
+	message := (*previousStageOutput).(map[any]any)["message"].(string)
+
+	s.message <- message
+}
+
+func TestProvider_Utility(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jaredoconnell) [reply internal/step/plugin/provider_test.go:109]: Moving this to a documented issue.

**Review by jaredoconnell — APPROVED** [review]
Looks good to me, except for the tests' documentation. I moved that to an issue: https://github.com/arcalot/arcaflow-engine/issues/96

**Pull request merged**